System default theme added Fixes #1397

4shutosh opened the pull request:
<!--
Note: This checklist is a reminder of our shared engineering expectations. Feel free to change it, although assigning a GitHub reviewer and the items in bold are required.
-->

Task/Issue URL: #1397

**Description**:
Added system default theme option for supporting devices Android 10 onwards

**Steps to test this PR**:
1. Open Settings
2. Open theme group
3. Devices above 10 and above will have an option to set the System Default theme option.

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

4shutosh commented:
Doubts: 
- How do I add all the string values of other language translations, is there any efficient option other than adding them manually?
- Will need the Pixel param for the new SYSTEM_DEFAULT click statistics `AppPixelName:SETTINGS_THEME_TOGGLED_SYSTEM_DEFAULT`

4shutosh commented:
I've added translations as well.

Here is a screenshot of the entire implementation:
![sampleSS](https://user-images.githubusercontent.com/53831987/131323961-c4d74705-8e7b-4d21-b6b8-c12edecae736.jpeg)

Let me know with any changes, Please Review :)

4shutosh commented:
Video Sample for the same:
![videoSample](https://user-images.githubusercontent.com/53831987/131325237-0575e953-e4c4-47a1-b4d6-1e4dccb44cb9.gif)

malmstein commented:
@4shutosh thanks for the contribution! We are discussing internally how we want to present the Light / Dark / System options, but it's looking good so far. 

I'll get back to you with the next steps regarding translations, etc...

Good work!

4shutosh commented:
Sure @malmstein, please let me know of any changes required.

malmstein commented on app/src/main/java/com/duckduckgo/app/pixels/AppPixelName.kt:
```diff
@@ -111,6 +111,7 @@ enum class AppPixelName(override val pixelName: String) : Pixel.PixelName {
     SETTINGS_OPENED("ms"),
     SETTINGS_THEME_TOGGLED_LIGHT("ms_tl"),
     SETTINGS_THEME_TOGGLED_DARK("ms_td"),
+    SETTINGS_THEME_TOGGLED_SYSTEM_DEFAULT("ms_tsd"), // confirm? the code here for statistics
```
this is good, just remove the comment

malmstein commented on app/src/main/res/layout/content_settings_general.xml:
```diff
@@ -29,14 +29,15 @@
-        android:text="@string/settingsLightTheme"
-        android:theme="@style/SettingsSwitchTheme"
+    <com.duckduckgo.app.settings.SettingsOptionWithSubtitle
+        android:id="@+id/themeToggle"
```
can you rename this to `selectedThemeSetting` to follow convention? It's not really a toggle anymore.

malmstein commented on app/src/main/res/values/strings.xml:
```diff
@@ -151,6 +151,9 @@
     <string name="settingsHeadingOther">Other</string>
     <string name="settingsHeadingPrivacy">Privacy</string>
     <string name="settingsLightTheme">Light Theme</string>
+    <string name="settingsTheme">Theme</string>
```
new translations go into `settings-untranslated.xml` and only in English

malmstein commented on app/src/main/res/values/strings.xml:
```diff
@@ -151,6 +151,9 @@
     <string name="settingsHeadingPrivacy">Privacy</string>
     <string name="settingsLightTheme">Light Theme</string>
+    <string name="settingsTheme">Theme</string>
+    <string name="settingsDarkTheme">Dark Theme</string>
```
ditto, also let's just have `Dark` and remove `Theme`

malmstein commented on app/src/main/res/values/strings.xml:
```diff
@@ -151,6 +151,9 @@
     <string name="settingsLightTheme">Light Theme</string>
+    <string name="settingsTheme">Theme</string>
+    <string name="settingsDarkTheme">Dark Theme</string>
+    <string name="settingsSystemTheme">System Default</string>
```
ditto

malmstein commented on app/src/main/res/values/strings.xml:
```diff
@@ -493,6 +496,9 @@
     <string name="disableLoginDetectionDialogNegative">CANCEL</string>
     <string name="fireproofWebsiteRemovalConfirmation">Fireproofing removed for &lt;b>%s&lt;/b></string>
 
+    <!--    Theme Settings-->
```
ditto

malmstein commented on app/src/main/res/values/strings.xml:
```diff
@@ -493,6 +496,9 @@
     <string name="fireproofWebsiteRemovalConfirmation">Fireproofing removed for &lt;b>%s&lt;/b></string>
 
+    <!--    Theme Settings-->
+    <string name="settingsThemeDialogSave">Set Theme</string>
```
ditto

malmstein commented on app/src/main/res/layout/settings_theme_selector_fragment.xml:
```diff
@@ -0,0 +1,68 @@
+            android:text="@string/settingsDarkTheme"/>
+
+        <RadioButton
+            android:id="@+id/themeSelectorSystemDefault"
```
can you move System Default to be the first item in the dialog?

malmstein commented on common-ui/src/main/java/com/duckduckgo/mobile/android/ui/Theming.kt:
```diff
@@ -30,7 +33,8 @@ import com.duckduckgo.mobile.android.ui.Theming.Constants.THEME_MAP
     DARK,
-    LIGHT;
+    LIGHT,
+    SYSTEM_DEFAULT;
```
`SYSTEM_DEFAULT` should be the first in the list

malmstein commented on app/src/main/res/values/strings.xml:
```diff
@@ -151,6 +151,9 @@
     <string name="settingsHeadingOther">Other</string>
     <string name="settingsHeadingPrivacy">Privacy</string>
     <string name="settingsLightTheme">Light Theme</string>
```
Can you rename this to only `Light`?

malmstein commented on app/src/main/java/com/duckduckgo/app/settings/SettingsThemeSelectorFragment.kt:
```diff
@@ -0,0 +1,100 @@
+    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
+
+        val currentOption: DuckDuckGoTheme =
+            arguments?.getSerializable(DEFAULT_OPTION_EXTRA) as DuckDuckGoTheme? ?: SYSTEM_DEFAULT
```
the default here should be `LIGHT`

malmstein reviewed (changes requested):
Thanks for this @4shutosh!

I've left a few comments, but mostly:

- Delete all the new added translations to different languages. We have an internal process to deal with them
- Move the new added English strings to `strings-untranslated.xml`
- Change order of dialog so System Default is the first item on the list.

4shutosh commented:
On it

4shutosh commented:
- I've removed translations, also moved `settingsLightTheme` string into the untranslated as the string value was changed to "Theme"
- Made System Default on the top of the list

malmstein commented on app/src/main/res/values/string-untranslated.xml:
```diff
@@ -40,4 +40,13 @@
+    <!-- Settings Activity -->
+    <string name="settingsLightTheme">Light</string>
+    <string name="settingsTheme">Theme</string>
+    <string name="settingsDarkTheme">Dark </string>
```
trailing spance

malmstein commented:
@4shutosh I left a comment about a space added that's not needed. Also, seems like there are some compilation errors:

```
> Task :app:compilePlayDebugAndroidTestKotlin FAILED
e: /Users/runner/work/Android/Android/app/src/androidTest/java/com/duckduckgo/app/settings/SettingsViewModelTest.kt: (148, 20): Unresolved reference: onLightThemeToggled
e: /Users/runner/work/Android/Android/app/src/androidTest/java/com/duckduckgo/app/settings/SettingsViewModelTest.kt: (159, 16): Unresolved reference: onLightThemeToggled
e: /Users/runner/work/Android/Android/app/src/androidTest/java/com/duckduckgo/app/settings/SettingsViewModelTest.kt: (167, 20): Unresolved reference: onLightThemeToggled
e: /Users/runner/work/Android/Android/app/src/androidTest/java/com/duckduckgo/app/settings/SettingsViewModelTest.kt: (179, 16): Unresolved reference: onLightThemeToggled
e: /Users/runner/work/Android/Android/app/src/androidTest/java/com/duckduckgo/app/settings/SettingsViewModelTest.kt: (187, 20): Unresolved reference: onLightThemeToggled
```

Once those are fixed we are ready to approve. Let me know if you need any help!

4shutosh commented:
@malmstein I've modified the tests. Please check

I've noticed one thing while implementing this that the `DialogFragments` do not follow the dark theme for some reason. I would like to open an issue for this and work on the same.

4shutosh commented:
Another doubt: There is a `darkThemeSwitch` inside AppComponentsActivity inside the Internal Features, does that require any modification as well?

4shutosh commented:
Is there anything else required to be done from my side @malmstein? Please let me know

malmstein commented:
It's all good now @4shutosh. Thanks for putting in the work!

4shutosh commented:
It's my pleasure. Thank you for you help 🙌🏻
